torch/utils/data/dataloader.py

```diff
@@ -352,7 +352,11 @@ def _next_data(self):
 
     def __next__(self):
         data = self._next_data()
-        self._num_yielded += 1
+        # num_yielded and the value of _IterableDataset_len_called are in units of "samples" rather than
+        # batches, so when using auto_collation, we need to increment by the length of the batch we're
+        # yielding
+        self._num_yielded += len(data) if self._auto_collation else 1
+
```

Comment:
also worth adding a comment that things like `drop_last` will not be falsely triggering the following check.